PERF Shrink arrays to size in liblinear helper dense_to_sparse

alexhenrie opened the pull request:
Calling realloc to shrink the size of a previously allocated array is
guaranteed to succeed. This way is faster because the array typically
does not need to be copied to a new location, it requires less memory
because there is no temp array that has to coexist with the last array
to be allocated, and it eliminates the need for gotos.

rth commented on sklearn/svm/src/liblinear/liblinear_helper.c:
```diff
@@ -20,20 +20,23 @@ static struct feature_node **dense_to_sparse(double *x, npy_intp *dims,
 {
```
The function docstring above would also need to be updated I think,
> We create a temporary array temp that collects non-zero
> elements and after we just memcpy that to the proper array.

alexhenrie replied:
Good catch! I just force-pushed a new commit with a corrected docstring.

rth reviewed (commented):
Thanks @alexhenrie ! Would you have an estimation of how much this improves performance (roughly)?

alexhenrie commented:
I don't have a great setup for taking a benchmark, but I will try to produce one if you need it. In my case, our memory is so scarce and our storage so slow that anything that helps avoid using swap space is significant for us.

rth commented:
> I don't have a great setup for taking a benchmark, but I will try to produce one if you need it

No worries I can quickly check it tomorrow. It was mostly to determine what to write in the "what's new" entry (and if one was needed).

jnothman reviewed (approved):
> It was mostly to determine what to write in the "what's new" entry (and if one was needed).

Yes, I wondered the same.

alexhenrie committed:
PERF Shrink arrays to size in liblinear helper dense_to_sparse

Calling realloc to shrink the size of a previously allocated array is
guaranteed to succeed. This way is faster because the array typically
does not need to be copied to a new location, it requires less memory
because there is no temp array that has to coexist with the last array
to be allocated, and it eliminates the need for gotos.

alexhenrie commented:
@rth Could you just commit this? You really don't have to mention it in the changelog.

rth commented:
Regarding benchmarks, I can't really measure any impact of this change on `LogisticRegression(solver="liblinear", multi_class="ovr")` using dense data with `n_samples=750000, n_features=200` and the following benchmark script,
<details>

```py
from sklearn.linear_model import LogisticRegression
from sklearn.datasets import make_classification

from neurtu import delayed, Benchmark
from sklearn.model_selection import train_test_split


X, y = make_classification(
    n_samples=1000000, n_features=200, n_informative=150, n_classes=2,
    random_state=2
)

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2)


bench = delayed(LogisticRegression)(
    solver="liblinear", multi_class="ovr", random_state=2
).fit(X_train, y_train)

print(f"X_train.shape={X_train.shape} X_train.size={X_train.size/1e6:.2f} MB")

df = Benchmark(wall_time=True, cpu_time=True, peak_memory={"interval": 0.01}, repeat=3)(
    bench
)

print(df[["cpu_time", "wall_time", "peak_memory"]])

# check scores
est = LogisticRegression(solver="liblinear", multi_class="ovr", random_state=2)
est.fit(X_train, y_train)
print("Train score: ", est.score(X_train, y_train))

print("Test score: ", est.score(X_test, y_test))
```
</details>
both version return identical results up to the measurement precision,

```
X_train.shape=(750000, 200) X_train.size=150.00 MB
9it [04:18, 30.98s/it]
       cpu_time  wall_time  peak_memory
mean  32.304107  32.113483  2368.906250
max   32.331746  32.123570  2377.203125
std    0.031689   0.010082    14.336793
Train score:  0.8396866666666667
Test score:  0.838568
```
wall_time/cpu_time is in seconds and peak_memory is in MB (with a sampling rate of 10ms).

I have checked that the changed function was indeed called in this benchmark and the the resulting train/test scores haven't changed.

@alexhenrie Do you use it in different conditions where it matters more? I'm OK merging this as the code gets simpler, but I would still like to understand why I'm not seeing any performance improvement that you mentionned.

alexhenrie commented:
We do logistic regression over about 65 million samples of about 20 features each, but even so, it may be that the performance improvement from using realloc instead of memcpy is negligible compared to the time and memory it takes to do the logistic regression itself. Based on your benchmarks I'd say that these changes do not merit a changelog entry.

rth reviewed (approved):
OK, makes sense. Thanks @alexhenrie !

alexhenrie commented:
@rth Thank you!
